aspnetcore/test/troubleshoot-azure-iis.md

```diff
@@ -0,0 +1,607 @@
+
+### Obtain data from an app
+
+If an app is capable of responding to requests, obtain request, connection, and additional data from the app using terminal inline middleware. For more information and sample code, see <xref:test/troubleshoot#obtain-data-from-an-app>.
+
+### Slow or hanging app (IIS)
+
+A *crash dump* is a snapshot of the system's memory and can help determine the cause of an app crash, startup failure, or slow app.
+
+#### App crashes or encounters an exception
+
+Obtain and analyze a dump from [Windows Error Reporting (WER)](/windows/desktop/wer/windows-error-reporting):
+
+1. Create a folder to hold crash dump files at `c:\dumps`. The app pool must have write access to the folder.
+1. Run the [EnableDumps PowerShell script](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/test/troubleshoot-azure-iis/scripts/EnableDumps.ps1):
+   * If the app uses the [in-process hosting model](xref:host-and-deploy/iis/index#in-process-hosting-model), run the script for *w3wp.exe*:
+
+     ```console
+     .\EnableDumps w3wp.exe c:\dumps
+     ```
+
+   * If the app uses the [out-of-process hosting model](xref:host-and-deploy/iis/index#out-of-process-hosting-model), run the script for *dotnet.exe*:
+
+     ```console
+     .\EnableDumps dotnet.exe c:\dumps
+     ```
+
+1. Run the app under the conditions that cause the crash to occur.
+1. After the crash has occurred, run the [DisableDumps PowerShell script](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/test/troubleshoot-azure-iis/scripts/DisableDumps.ps1):
+   * If the app uses the [in-process hosting model](xref:host-and-deploy/iis/index#in-process-hosting-model), run the script for *w3wp.exe*:
+
+     ```console
+     .\DisableDumps w3wp.exe
+     ```
+
+   * If the app uses the [out-of-process hosting model](xref:host-and-deploy/iis/index#out-of-process-hosting-model), run the script for *dotnet.exe*:
+
+     ```console
+     .\DisableDumps dotnet.exe
+     ```
+
+After an app crashes and dump collection is complete, the app is allowed to terminate normally. The PowerShell script configures WER to collect up to five dumps per app.
+
+> [!WARNING]
+> Crash dumps might take up a large amount of disk space (up to several gigabytes each).
+
+#### App hangs, fails during startup, or runs normally
+
+When an app *hangs* (stops responding but doesn't crash), fails during startup, or runs normally, see [User-Mode Dump Files: Choosing the Best Tool](/windows-hardware/drivers/debugger/user-mode-dump-files#choosing-the-best-tool) to select an appropriate tool to produce the dump.
+
+#### Analyze the dump
+
+A dump can be analyzed using several approaches. For more information, see [Analyzing a User-Mode Dump File](/windows-hardware/drivers/debugger/analyzing-a-user-mode-dump-file).
+
+## Additional advice
```

Comment:
Since this is about clearing the package caches, would "Clear package caches" be a better h2?

Reply:
I couldn't do it because there are several *different* instructions here. I'm glad you said something tho. I don't like the composition of this little section:

* It's not a step-wise list of ordered troubleshooting items. It's just a plain old list.
* The package cache clearing "TIP" (CLI command) really should be in the instruction.

... so I'm going to make it into a set (not steps) of advice items ...

* Delete the *bin* and *obj* folders.
* Clear the package caches by executing `dotnet nuget locals all --clear` from a command shell.

  Clearing package caches can also be accomplished by using the [nuget.exe](https://www.nuget.org/downloads) tool and executing the command `nuget locals all -clear`. *nuget.exe* isn't a bundled install with the Windows desktop operating system and must be obtained separately from the [NuGet website](https://www.nuget.org/downloads).

* Restore and rebuild the project.
* Confirm that the prior deployment on the server has been completely deleted prior to redeploying the app.